### ISSUE TITLE
Deep packet inspection to classify V2Ray traffic

### ISSUE THREAD

**Issue by rickyzhang82** [v2ray/discussion#569]
In last 3 months, I'm verifying  [the patent claim](https://files.catbox.moe/vmzj04.pdf) by replicating it with code and data.

First, I spent two months to write a [TCP packet sorter](https://github.com/rickyzhang82/PacketSorter).  Then I collected home traffic for a month where I mixed v2ray traffic with other normal network traffic.

Last night, I stopped capturing packet data and started to build a classifier. I spent two hours to write a 1D Convolution Neutral Network to do binary classification on v2ray traffic and non-v2ray traffic. 

To my surprise, I can achieve classification accuracy rate up to 98.86%. In a nutshell, the patent claim is real. 

I shared my [classification model in Python notebook](https://github.com/rickyzhang82/V2Ray-Deep-Packet-Inspection/blob/master/V2Ray%20Deep%20Packet%20Inspection.ipynb) and findings in [my blog post](https://fr33land.net/2020/01/18/we-are-fucked/).

**Comment by rickyzhang82** [comment]
The issue is crystal clear. The vmess protocol over TCP can be identified without using decryption, active probe, port or IP. Thus, it can be blocked in practice. I read [the research report on shadownsocks](https://gfw.report/blog/gfw_shadowsocks/). If the tool can be detected, the next step is to be blocked by GFW. Don't we learn the lesson now?

My initial proposal is to disable the settings using the vmess protocol over TCP out of the box. It is unwise to expose citizens at risk if we know for sure that it can be identified. Can V2Ray dev team make this change? Or at least update your official user guide to inform users about this.

I'm still collecting data with the setting using vmess over TLS, where traffic is masqueraded as TLS. I'm not sure this is a solution. Because we can classify SMTP over VPN, web traffic over VPN, and etc. Why TLS can make a difference? This is my speculation. It takes time to prove it.

I don't use V2Ray. I don't need any circumvention tools to access Internet. The only reason I spent my time here is to defend freedom of expression.

**Comment by nicholascw** [comment]
What is crystal clear is that your logic is problematic. Disabling VMess over TCP is exact same logic as the case below, which is ridiculous:
```
   New study shows that there're measures to poison water without noticed by target victim.
=> There must be people poison any target water source that use water for whatever reasons.
=> We should cease water supply to anybody as soon as possible.
```
Though V2Ray project identifies itself as a "unified platform for anti-censorship." which means it takes anti-censorship in high priority, however the prerequisite is it is still a unified platform, in other words, a versatile network utility. As a core function and foundation of the whole project, I don't think there will be any chance we stop supporting VMess over TCP in any near future.

With basic statistics knowledge you should quickly realize that you won't gain a bigger possibility of successfully pass a firewall with an option disabled.

**Comment by nicholascw** [comment]
What is crystal clear is that your logic is problematic. Disabling VMess over TCP is exact same logic as the case below, which is ridiculous:
```
   New study shows that there're measures to poison water without noticed by target victim.
=> There must be people poison any target water source that use water for whatever reasons.
=> We should cease water supply to anybody as soon as possible.
```
Though V2Ray project identifies itself as a "unified platform for anti-censorship." which means it takes anti-censorship in high priority, however the prerequisite is it is still a unified platform, in other words, a versatile network utility. As a core function and foundation of the whole project, I don't think there will be any chance we stop supporting VMess over TCP in any near future.

With basic statistics knowledge you should quickly realize that you won't gain a bigger possibility of successfully passing a firewall with an option disabled.

> If you don't like it, fork it. [1]
```
[1] Zhang, R. @rickyzhang82 GitHub Profile. Retrieved February 27, 2020, from https://github.com/rickyzhang82
```

**Comment by rickyzhang82** [comment]
Are we playing a GRE analytical writing game here? If so, I will point you out the fallacy in your analogy.

There are many ways to deliver safe water. If we know for sure that water delivering through the city water pipe may contain poisonous lead, should you stop the water distribution this way? Hell, yes. Look up the news on "Flint Water Crisis".

But it doesn't mean we stop water supply completely. People in Flint, Michigan still need water. Therefore, they switched to drinking bottled water temporarily. At the same time, the city is rebuilding a lead-free water pipe system.

The same logic applies to V2Ray. There are many settings provided by V2Ray. I never advise people to discontinue V2Ray completely so that no one can circumvent GFW. I just propose to stop using a specific insecure configuration that has been proven or at least give a warning in the user guide.

My statistical background can **NOT** come to your conclusion. 

> With basic statistics knowledge you should quickly realize that you won't gain a bigger possibility of successfully passing a firewall with an option disabled.

If you imply the non-techie should use the insecure setting to increase the chance for the techie to evade GFW, that makes me sick.

You don't need to quote my words from my profile. This will be my next step.

**Comment by nicholascw** [comment]
1. You are substituting _water_ with _specific water source_.
  VMess itself which is a protocol, is generally the same as _water_ rather than _specific water source_ until you deploy it into a specific environment.

2. There are various ways provided to wrap, encapsulate, and deliver VMess including features like UNIX socket, mKCP, mux, websocket, tls. Features are provided and their working mechanism, properties and performance are described in each page of the official manual. We never tend to guide user to use ANY form of configurations into ANY environment. It is the users' job to decide which way is the way that fits their need.

3. At the same time, you shouldn't automatically assume in any degree that the provided configuration file sample, which is the most basic, most plain VMess configuration demo, is a recommended way to use in any circumstances. We made no recommendations in the document, neither VMess or any other combined configuration, and the demo files were their simply act as a skeleton for user to modify it. Also, there are more configuration templates to be found at https://github.com/KiriKira/vTemplate, which is a repo listed as a third party tool in the manual, not to mention that we also forked, maintained, and internationalizing the ToutyRater's V2Ray Guide and operate at https://guide.v2fly.org, which listed a bunch of configuration scenarios and intend to let "non-techie" understand. 

> THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

**Comment by vrnobody** [comment]
I am really interesting about the vmess-ws-tls combination.
Sadly i know nothing about machine learning, so i could not conduct the experiment myself. 

Have you finished the second-stage data collection? And what is the result?

**Comment by rickyzhang82** [comment]
@nicholascw 

> VMess itself which is a protocol, is generally the same as water rather than specific water source until you deploy it into a specific environment. Removing VMess over TCP is like forbidding use non-drinkable water to run a steam engine.

You reason like [a Chinese foreign ambassador in Aussie](https://youtu.be/FyDjYjcgkdo): 

- Evade the crux of the matter.
- Shift the focus like a monkey jumping here and there.

The vmess protocol is neither water nor water source. We are talking about a flawed application protocol that may expose the users' circumvention from GFW.

But if you want to play with the analogy, then fine. Let's do it **right**. Free information behind GFW is like the clean water we desperately want. The vmess protocol over TCP configuration in V2Ray is like the city water pipe that may leak the poisonous lead. If we knew this, we should inform the techies and also **the non-techies** about the risk of consumption. 

Can we find such **FYI precautions** in the user guide? No, it doesn't exist.

I'm not sure why you quote the limitation of liability clause at the end of your argument. Do you imply that it is a justification to allow the existence of a flawed application protocol without a warning? That's a silly argument.

@vrnobody 

I'm done with one-month-long data collecting with the TLS+WS setting yesterday. The DNN model can still achieve >99% classification accuracy in the validation set (I split 80% of data for the training set and 20% for the validation set.)

In the following weeks, I will collect the traffic data from a different host with a different TLS certificate and a different UUID and see if the previously trained model can generalize classification capabilities. 

Once all are done, I will release a blog post in public.

**Comment by vrnobody** [comment]
Oh no! Basically that means we are doomed!

Thanks for your hard work.

**Comment by rickyzhang82** [comment]
@madeye @vrnobody 

I published my finding on TLS + WS in [my blog post](https://fr33land.net/2020/03/12/can-enable-tls-in-v2ray-help/) and uploaded pre-trained model as well.

In a nutshell, the pre-trained model can generalize its classification capabilities to different UUID, different IP, different ports, different domain names, different OS flavors, and different TLS certificates.

Can @madeye elaborate "obfuscating" solutoin?
